staging/src/k8s.io/client-go/tools/events/event_recorder.go

```diff
@@ -45,19 +45,20 @@ func (recorder *recorderImpl) Eventf(regarding runtime.Object, related runtime.O
 	message := fmt.Sprintf(note, args...)
 	refRegarding, err := reference.GetReference(recorder.scheme, regarding)
 	if err != nil {
-		klog.Errorf("Could not construct reference to: '%#v' due to: '%v'. Will not report event: '%v' '%v' '%v'", regarding, err, eventtype, reason, message)
+		// TODO: extend Eventf with ctx parameter.
```

Comment:
There are currently 226 such calls. That's a lot, but IMHO it is the right thing to do. Otherwise the error will not get logged as part of the current activity which triggered the event.

I prefer `context` over `klog.Logger` here because if we ever need to do some real work inside `Eventf`, that context would provide the deadline for it.

Reply:
Also an important point in favor of context: the logger is normally not needed, so the usual argument against `context` when a logger is needed (the extra lookup) doesn't apply.

Reply:
I wonder if it is really worth adding a context to `Eventf`. We have 226 such calls in-tree, but it will also impact the consumers of client-go. Realistically, this change will impact the majority of the operators that are in the wild today.

> here because if we ever need to do some real work inside Eventf

I don't think we will ever need to do anything more than create an Event object here. All the aggregation and apiserver related logics are done at the broadcaster's level today and I don't think this is bound to change.
Also, with the changes you made from this PR, we will now be able to cancel broadcasting from the context, which isn't something we could do before.

Reply:
The alternative is to add an `EventfWithLogger` which takes a `klog.Logger` as first parameter. Using that function is optional, i.e. `Eventf` will not get deprecated, but code which cares about getting the error log messages recorded "correctly" have an API to do it.

Because it's not a context, `Eventf` can be implemented as `EventfWithLogger(klog.Background(), ...)`. If we needed a context, `Eventf` would remain incomplete because it would forever have to use `context.TODO`.

Reply:
I tried a different approach: `EventRecorder.WithContext` creates a new EventRecorder which holds a context for the `Eventf` call. I chose that approach because `record.EventRecorder` contains three calls which all would have needed separate variants with an explicit `ctx` parameter.

Extending a published interface that could have been implemented elsewhere is always risky because it is an API break. I think it's safe in this case.

@dgrisonnet: what do you think?

Reply:
Change of plan... there are enough "fake EventSource" implementations that extending that interface is annoying. I changed the code so that `EventSource` is unchanged and added `EventSourceContext` as extension of it. To avoid type casting, constructors return that. This leads to an API change in `EventBroadcaster` but perhaps that one is only used and not implemented by clients?

Code using the API still works as before:
```
var recorder EventRecorder

recorder := broadcaster.NewRecorder(...)
```

Reply:
Passing an interface instead of plain `ctx` or `logger` turned out to be too complicated. As some code only has a logger and changing that code would be more complicated than adding `klog.FromContext` in code which has a `ctx`, let's just use `logger`.

Reply:
I've pushed the version which uses `WithLogger` instead of `WithContext`.

Converting existing code is often as easy as `s/recorder record.EventRecorder/recorder record.EventRecorderLogger/` and `s/recorder.Event/recorder.WithLogger(logger).Event/`. I converted two controllers as a trial, but won't include those changes in this PR - that can come separately.

Reply:
I am fine with your last approach with `WithLogger`. It feels a bit odd to have the logger in a context in the broadcaster and a logger directly in the recorder, but since the recorder doesn't need a context, I think it makes sense.